soccer/gameplay/test_list.py

```diff
@@ -0,0 +1,91 @@
+from PyQt5 import QtCore
+import test_registry
+from enum import Enum
+
+
+class Status(Enum):
+    idle = 0
+    running = 1
+    completed = 2
+
+
+class TestNode(test_registry.TestRegistry.Node):
+    def __init__(self, registryNode):
+        super().__init__(registryNode._module_name, registryNode.test_class)
+        self.reset()
+
+    def reset(self):
+        self.test = self.test_class()
+        self.status = Status.idle
+        self.results = []
+
+
+class TestList(QtCore.QAbstractListModel):
+    def __init__(self, selectedTestsTable):
+        super().__init__()
+        self.tList = []
+        self.selectedTestsTable = selectedTestsTable
+
+    def insert(self, item):
+        self.tList.append(TestNode(item))
+        index = self.createIndex(len(self.tList) - 1, 0)
+        self.dataChanged.emit(index, index)
+
+    def remove(self, node):
+        i = self.tList.index(node)
+        del self.tList[i]
+
+        index = self.createIndex(i, i)
+
+        # After a lot of experimentation this is the closest that I
+        # can get to the correct implementation of this.
+        self.dataChanged.emit(index, index)
+        self.beginRemoveRows(self.createIndex(0, 0), i, i)
+        self.rowsRemoved.emit(self.createIndex(0, 0), i, i)
+        self.endRemoveRows()
+
+        return i
+
+    def rowCount(self, parent):
+        return len(self.tList)
+
+    def data(self, index, role=None):
+        if not index.isValid():
+            return None
+        node = index.internalPointer()
+
+        # This if is annoyingly necessary
+        # QT somehow calls this on garbage data
+        # during the first iteration after a remove
```

Comment:
...excuse me what?

Okay sure seems reasonable, have a nice day.

Reply:
Something with the way QT handles the resize in the qAbstractList causes it to try to index role_requirements objects and a few other weird things through the data function. I did some testing but I was never able to figure out why :man_shrugging: